src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py

```diff
@@ -13,6 +15,97 @@
 from ...optimizers.qaoa_optimizer import get_optimizer
 
 
+def circuit_routing(
+    device: DeviceBase,
+    problem_to_solve: List[List[int]],
+    initial_mapping: Optional[List[int]] = None,
+    routing_algo: str = "greedy",
+) -> tuple:
+    """Post the HTTP request to the routing API and retrieve routed
+    circuit
+
+    Parameters
+    ----------
+    device: DeviceBase
+        The device for which to route the circuit
+    cost_block: List[GateMap]
+        The QUBO problem to be run on QPU
+    initial_mapping: List[int]
+        The initial selection of qubits, if provided by the user.
+        Defaults to the first n-qubits on the device if `None`
+    routing_algo: str
+        The algorithm to generate an optimal SWAP network
+
+    Returns
+    -------
+    tuple[List[List],List[bool],List[int]]
+        The first list specifies the qubit indices for operations,
+        the second list differentiates between Ising and SWAP gates,
+        and the final list returns the final qubit mapping.
+    """
+    device_connecivity = device.connectivity()
+
+    # params for MCTS
+    params = (
+        {
+            "max_depth_simulation": 10,
+            "max_iterations": 10,
+            "n_simulations": 64,
+            "exploration_constant": 2,
+            "max_tree_depth": 5,
+        }
+        if routing_algo.lower() == "mcts"
+        else {}
+    )
+    query = {
+        "algorithm": routing_algo,
+        "device": device_connecivity,
+        "problem": problem_to_solve,
+        "initial_mapping": initial_mapping,
+        "params": params,
+    }
+
+    # post the query via HTTP
+    r = requests.post(url="http://127.0.0.1:8000/qubit_routing", json=query)
+
+    results = r.json()
+
+    gate_indices_list = results["gates_list"]
+    initial_mapping = results["initial_mapping"]
+    final_mapping = results["final_mapping"]
+    swap_mask = results["swap_mask"]
+    
+    all_qubits = []
+    for pair in gate_indices_list:
+        all_qubits.extend(pair)
+    all_qubits = set(all_qubits)
+    
+    #append missing qubits from the initial_mapping
+    initial_mapping.extend([qubit for qubit in all_qubits if qubit not in initial_mapping])
+    #append missing qubits from the final_mapping
+    final_mapping.extend([qubit for qubit in all_qubits if qubit not in final_mapping])
+    
+    initial_physical_to_logical_mapping = {
+        qubit:idx for idx,qubit in enumerate(initial_mapping)
+    }
+    
+    for idx,pair in enumerate(gate_indices_list):
+        i,j = pair
+        logical_i, logical_j = initial_physical_to_logical_mapping[i],initial_physical_to_logical_mapping[j] 
+        gate_indices_list[idx] = sorted([logical_i, logical_j])
```

Comment:
For ease of use within OQ, we always order qubits in increasing order. Since QR returns a list of gates (defined with their indices), and those gates are either RZZ or SWAP (both symmetric in qubits), their order does not matter physically. The decision to sort the indices is entirely software in nature.

Reply:
Ok sounds good.